### ISSUE TITLE
align current rust module types with new requirements

### ISSUE THREAD

**Issue by ejmg** [amadeusine/CS449GroupProject#23]
*this issue is included in the #22 tracker for Sprint 2 milestone*

---

# Summary

Currently, there is now some incongruity between the types defined by the rust module and those needed by our requirements as previously discussed in meetings. The implemented front-end only mocks rendering logic and doesn't have any equivalents yet, so now is the time to fix this before confusion arises.

As of now, we need to do the following additions:
- [x] `GameState`
  - [x] add `Handle` type
     - [x] `Handle` is an enum with variants representing:
       - [x] "Ok", i.e. whether last attempted move was successful.
       - [x] "Err i.e. the last attempted move failed.
  - [x] Add `Trigger` type
    - [x] `Trigger` type is an enum with variants:
       - [x] `None`
       - [x] varients representing each stage of game:
          - [x] Start of game
          - [x] End of game
          - [x] Player selection
          - [x] Placing pieces
          - [x] Elimination
          - [x] Flying

## clarifications

- `GameState`
  - meeting minutes (and discussion itself) focused around `BoardStruct`, when previous meeting minutes and design discussions has this as `GameState`, which is also what the rust module uses as the name for this structure.
    - raises general issue of whether we need to split it into two types, since currently it holds `Board` and `Mill` types, which represent the next board layout and the set of existing mills, respectively.
    - previous meeting minutes we had notion of another `State` type which returned:
      - "`Success` for `InputHandle`"
      - "`BoardStruct` is the new `State`"
      - "`Trigger" of the condition of the phase is satisfied'
      - **do we still want this?**
  - `Trigger` type
      - Given `None` variant, maybe the better/intended program semantic is to make the type itself an Option?
      - Could the notion of Start and end game phases be better encapsulated into one variant which represents the general notion, like "menu"?

##  potential blockers

None

**Comment by ejmg** [comment]
Conversation and stub code brainstormed by @amadeusine is changing `GameState` towards the following definition:
``` rust
struct GameState {
    handle: Handle,
    trigger: Trigger,
    board: Board,
    mills: Vec<Mill>,
}
```

I've changed Board to be a hashmap of tuples, basically, as well:
``` rust
#[derive(Debug, Clone)]
struct Board(HashMap<Coord, PositionStatus>);
```
where `PositionStatus` is just a new type alias:
```rust
type PositionStatus = (bool, Option<Player>);
```

Finally, I've rethought how to represent a Mill in terms of how we would end up touching it, and I think the following declaration makes it less... Painful:
```rust
#[derive(Debug, Clone, PartialEq, PartialOrd)]
struct Mill {
    Owner: Player,
    Pieces: (Position, Position, Position),
}
```

**Comment by ejmg** [comment]
Adding `Manager` as an official type. It's been discussed a few times and escaped being implemented sooner somehow.

`Manager` is an a rust struct that gets JS'd to return the result of attempted actions from a user.

**Comment by ejmg** [comment]
Adding `GameOpts` as an official type that contains the meta-data related to a move being attempted from the front-end. Everything is being set as optional for now until I figure out the best way to take the json/js object passed from the front and convert that into rust types for acting upon.

```rust

enum Agent {
    Human,
    Auto,
}

pub struct GameOpts {
    user: Option<Player>,
    opponent: Option<Player>,
    agent: Option<Agent>,
    sender: Option<Player>,
    position: Option<Coord>,
}
```

As fleshed out thus far, the idea is there are two cases that define the contents actually provided within the `GameOpt` object. When the action being performed is menu related, i.e. starting a game, then `GameOpt` will have just:

```rust
struct GameOpt {
    user: Player,
    opponent: Player,
    agent: Agent,
}
```

When it's an action during any other part of the game, `GameOpt` would become:
```rust
struct GameOpt {
    sender: Player,
    position: Coord,
}

**Comment by ejmg** [comment]
`InputHandle` is getting reduced into a enum with `Ok` and `Err` variants to signal to the front-end whether the last attempted move was actually valid.

**Comment by ejmg** [comment]
Resolved with #29.
